src/stream/src/common/builder.rs

```diff
@@ -57,11 +68,43 @@ impl StreamChunkBuilder {
         Ok(Self {
             ops,
             column_builders,
+            data_types: data_types.to_owned(),
             update_start_pos,
             matched_start_pos,
+            capacity,
+            size: 0,
         })
     }
 
+    /// Append a row with coming update value and matched value, while tracking the chunk size under
+    /// the capacity limit
+    ///
+    /// A [`StreamChunk`] will be returned when `size == capacity`
+    pub fn append_row_with_limit(
+        &mut self,
+        op: Op,
+        row_update: &RowRef<'_>,
+        row_matched: &Row,
+    ) -> Result<Option<StreamChunk>> {
+        self.ops.push(op);
```

Comment:
Data chunks must not split update-insert and update-delete apart, so this doesn't seem like an ultimate solution for split chunk. We can revisit this part later.

Reply:
This may lead to unstable e2e tests. 🤤